synapse/http/federation/matrix_federation_agent.py

```diff
@@ -273,9 +317,42 @@ async def _do_connect(self, protocol_factory: IProtocolFactory) -> None:
             host = server.host
             port = server.port
 
+            should_skip_proxy = False
+            if self.no_proxy is not None:
+                should_skip_proxy = proxy_bypass_environment(
+                    host.decode(),
+                    proxies={"no": self.no_proxy},
+                )
+
+            endpoint: IStreamClientEndpoint
             try:
-                logger.debug("Connecting to %s:%i", host.decode("ascii"), port)
-                endpoint = HostnameEndpoint(self._reactor, host, port)
+                if self.https_proxy_endpoint and not should_skip_proxy:
+                    logger.debug(
+                        "Connecting to %s:%i via %s",
+                        host.decode("ascii"),
+                        port,
+                        self.https_proxy_endpoint,
+                    )
+                    connect_headers = Headers()
+                    # Determine whether we need to set Proxy-Authorization headers
+                    if self.https_proxy_creds:
+                        # Set a Proxy-Authorization header
+                        connect_headers.addRawHeader(
+                            b"Proxy-Authorization",
+                            self.https_proxy_creds.as_proxy_authorization_value(),
+                        )
```

Comment:
This change was larger.
I had to move `ProxyCredentials` from `proxyagent` to `connectproxyclient`. The reason was a circular import.
I have replaced `headers` parameter because it was introduced in #9657 only for proxy connections and is not needed anymore.